[fparser] New type definition

andrew0687 opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

andrew0687 reviewed (commented):
Ready to merge

andrew0687 commented:
Ready to merge

peterblazejewicz commented:
@andrew0687 IMO one should not be listed as owner, unless explicitly confirms it's OK to use relevant github handle. You would need that other person review here

peterblazejewicz commented on types/fparser/index.d.ts:
```diff
@@ -0,0 +1,28 @@
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export default class Formula {
+    constructor(fStr: string, options?: object);
```
the options are as simple as:
```suggestion
    constructor(fStr: string, options?: { memoization?: boolean | undefined });
```

peterblazejewicz commented on types/fparser/index.d.ts:
```diff
@@ -0,0 +1,28 @@
+    storeInMemory(valueObj: object, value: any): void;
+    getExpression(): any[];
+    getExpressionString(): string;
+    static calc(formula: string, valueObj: object | object[], options?: object): any;
```
```suggestion
    static calc(formula: string, valueObj: object | object[], options?: { memoization?: boolean | undefined; }): any;
```

peterblazejewicz commented on types/fparser/tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6",
+            "dom"
```
```suggestion
            "es6"
```
where is 'dom' reference required?

peterblazejewicz commented on types/fparser/index.d.ts:
```diff
@@ -0,0 +1,28 @@
+    splitFunctionParams(toSplit: string): string[];
+    cleanupInputString(s: string): string;
+    parse(str: string): any;
+    private _do_parse(str: string): any;
```
```suggestion
```
is this for internal use or for consumption in other code? If the last one, do not use `private`

peterblazejewicz commented:
as far as I can see, this is exported as UMD module, so it can be used as global script in a browser, is that correct?
https://www.typescriptlang.org/docs/handbook/modules.html#umd-modules

andrew0687 commented:
> as far as I can see, this is exported as UMD module, so it can be used as global script in a browser, is that correct? https://www.typescriptlang.org/docs/handbook/modules.html#umd-modules

No. Package used as node module.

peterblazejewicz commented:
I understand, I meant: exported and published:
https://unpkg.com/fparser@2.0.1/dist/fparser.js
https://www.npmjs.com/package/fparser#usage
```ts
window.Formula.calc;(...)
new window.Formula(...)
```

andrew0687 commented:
> I understand, I meant: exported and published: https://unpkg.com/fparser@2.0.1/dist/fparser.js https://www.npmjs.com/package/fparser#usage
> 
> ```ts
> window.Formula.calc;(...)
> new window.Formula(...)
> ```

Updated. Added declaration and test for UMD module

peterblazejewicz commented on types/fparser/test/fparser-tests.ts:
```diff
@@ -0,0 +1,7 @@
+import Formula from "fparser";
```
```suggestion
import Formula = require('fparser');
```
this would work, once you update `index.d.ts`.
`import Formula from 'fparser'` 'will work when consuming 'fparser' and types, when using modern TS config defaults, which is not a default setup on DT repo, to promote wider interoperability

peterblazejewicz commented on types/fparser/index.d.ts:
```diff
@@ -0,0 +1,9 @@
+//                 andrew0687 <https://github.com/andrew0687>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+///<reference path="./fparser.d.ts"/>
+
+export default FormulaC;
```
```suggestion
import Formula from './src/fparser';
export = Formula;

export as namespace Formula;
```
you can find this pattern in UMD lib documentation, should cover all usage of `fparser`.

peterblazejewicz commented on types/fparser/fparser.d.ts:
```diff
@@ -0,0 +1,43 @@
+
+    static calc(formula: string, valueObj: object | object[], options?: { memoization?: boolean; }): any;
+}
+
+declare const Formula: typeof FormulaC;
```
```suggestion
```

peterblazejewicz commented on types/fparser/fparser.d.ts:
```diff
@@ -0,0 +1,43 @@
+declare class FormulaC {
```
```suggestion
export default class Formula {
```
and move this file into `./src/` subdirectory (it's defined in `package.json` ESM support

peterblazejewicz reviewed (approved):
LGTM!
@andrew0687 thx!

andrew0687 commented:
Ready to merge
